src/Devices/EthernetServer_TC.cpp

```diff
@@ -419,29 +429,58 @@ void EthernetServer_TC::sendHeadersWithSize(uint32_t size) {
   // blank line indicates end of headers
   client.write('\r');
   client.write('\n');
-  state = FINISHED;  // TODO: Why?! This is awkward when we want to send a body next.
 }
 
-// 303 response
-void EthernetServer_TC::sendRedirectHeaders() {
-  static const char response[] PROGMEM =
+void EthernetServer_TC::sendResponse(int code) {
+  static const char response_303[] PROGMEM =
       "HTTP/1.1 303 See Other\r\n"
       "Location: /api/1/display\r\n"
       "Access-Control-Allow-Origin: *\r\n"
       "\r\n";
-  char buffer[sizeof(response)];
-  strncpy_P(buffer, (PGM_P)response, sizeof(buffer));
-  client.write(buffer);
-  state = FINISHED;
-}
-
-// 400 response
-void EthernetServer_TC::sendBadRequestHeaders() {
-  char buffer[30];
-  static const char response[] PROGMEM = "HTTP/1.1 400 Bad Request\r\n\r\n";
-  strncpy_P(buffer, (PGM_P)response, sizeof(buffer));
+  static const char response_400[] PROGMEM =
+      "HTTP/1.1 400 Bad Request\r\n"
+      "\r\n";
+  static const char response_404[] PROGMEM =
+      "HTTP/1.1 404 Not Found\r\n"
+      "\r\n";
+  static const char response_405[] PROGMEM =
+      "HTTP/1.1 405 Method Not Allowed\r\n"
+      "Allow: GET, POST\r\n"
+      "\r\n";
+  static const char response_408[] PROGMEM =
+      "HTTP/1.1 408 Request Timeout\r\n"
+      "Connection: close\r\n"
+      "\r\n";
+  static const char response_500[] PROGMEM =
+      "HTTP/1.1 500 Internal Server Error\r\n"
+      "\r\n";
+  static const char response_501[] PROGMEM =
+      "HTTP/1.1 501 Not Implemented\r\n"
+      "\r\n";
+  char buffer[sizeof(response_303)];
+  switch (code) {
+    case HTTP_REDIRECT:
+      strncpy_P(buffer, (PGM_P)response_303, sizeof(buffer));
+      break;
+    case HTTP_BAD_REQUEST:
+      strncpy_P(buffer, (PGM_P)response_400, sizeof(buffer));
```

Comment:
Does this mean the copy will always copy the same size buffer? Basically defaulting to the largest response text. Does it matter there is blank space?

Reply:
The third parameter to `strncpy_P()` does not specify how much will be copied but the size of the target buffer (to avoid buffer overrun bugs). The amount copied will be up to the null terminator or the size of the buffer, whichever is less.